File: Grokking the coding interview/Merge Intervals/57. Insert Interval.py
# TC: O(N) SC: O(N)

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        i, start, end = 0, 0, 1
        output = []

        while i < len(intervals) and intervals[i][end] < newInterval[start]:
            output.append(intervals[i])
            i += 1
        
        while i < len(intervals) and intervals[i][start] <= newInterval[end]:
            newInterval[start] = min(newInterval[start], intervals[i][start])
            newInterval[end] = max(newInterval[end], intervals[i][end])
            i += 1
        
        output.append(newInterval)

        while i < len(intervals):
            output.append(intervals[i])
            i += 1

        return output

File: Grokking the coding interview/Merge Intervals/57. Insert Interval.js
/**
 *  TC: O(N) SC: O(N)
 */

var insert = function(intervals, newInterval) {
    let output = [];
    for(let i = 0; i < intervals.length; i++){
        // Not intersecting
        // interval < new interval
        if(intervals[i][0] > newInterval[1]){
            output.push(newInterval, ...intervals.splice(i));
            return output;
        }
        else if (intervals[i][1] < newInterval[0]){
            output.push(intervals[i])
        }
        
        // Intersecting
        else{
            newInterval = [Math.min(intervals[i][0], newInterval[0]), Math.max(intervals[i][1], newInterval[1])]
        }
    }
    
    output.push(newInterval);
    return output
};